### DISCUSSION TITLE
Allow for short hand preparations in ingredients list to not double count

### DISCUSSION THREAD

**Discussion by rykard95** [cooklang/spec#57] [Ideas]
i.e. 3 scallions, white parts sliced thin, green parts cut into 1-inch pieces.

As a user, it would be nice if I could specify how each ingredient is supposed to be prepared.

**Comment by dubadub** [comment]
@rykard95 there's a chance I misunderstood, so please correct me if I'm wrong. 

I had this before in spec:

_To modify the ingredient any other way, use parentheses._
 ```
 Add @onions{3}(finely chopped) to the mix...
 ```

After some testing, it turned out that it makes the overall cooking flow a bit convoluted. This modifier is an implicit step of the recipe. Imagine that we have this modified ingredient in the middle of a recipe. I have my frying pan full of things and next step is to add onions. But I suddenly realised that it's time to finely chop onions and I don't have time for that.

We can create these steps automatically and place them before everything else, but it may not be always the optimal way. 

What do you think @rykard95? 

PS
Here it was removed: https://github.com/cooklang/cooklang-org/commit/1bd2f54c336379da0c4287c598bca502b7cec28b#diff-03461d0dfb4f81c8ad324ec60ebee82c2c3707031b02224367f970063fc484eeL42-L46

**Comment by deathau** [comment]
A lot of recipes I've seen do list those instructions in the ingredients though.
E.g. "3 Onions, finely chopped"

**Comment by chrbala** [comment]
The "mise en place" way here is to prep all your ingredients ahead of time before cooking everything. The idea is that upfront prep will reduce the likelihood of errors while cooking, and will generally make the cooking itself more enjoyable.

There are certainly some cases where prep is more efficiently done later, such as when the ingredient isn't used until after a long-running async process (e.g. chopped chives added to a finished soup that has to cook for half an hour).

Getting the distinction between AOT and JIT for ingredient prep is not easy though, and probably requires a bunch of heuristics. Probably not complexity you want to push to cooklang interpreters.

**Comment by zenahirsch** [comment]
Adding my humble vote here to return to something like what you had earlier in the spec. I am in the process of converting several recipes to Cooklang, and I'm finding that I'm often having to do something like:

```cooklang
When the sauce has thickened, stir in the @chopped fresh basil{1/4%cup}.
```

When I go to make a grocery list, it lists "chopped fresh basil," but really I'd be purchasing fresh basil and chopping it myself. Nor do I want to add a whole extra step at the top of the recipe that tells the user to chop the basil... that just doesn't seem to match the generally accepted style of formatting recipes.

I tried doing:

```cooklang
When the sauce has thickened, stir in the @fresh basil{1/4%cup, chopped}.
```

But then I end up with an ingredient list looking like:

```
fresh basil                   0.25 cup, choppeds
```

Specifying the preparation of an individual ingredient in parens seems like a good solution.

**Comment by dubadub** [comment]
Relevant proposal for spec change https://github.com/cooklang/spec/pull/106.

**Comment by dubadub** [comment]
This is now part of the main specification, closing the discussion.
